### PR TITLE
Update selectNodeVersion.js to use WEBSITE_NPM_DEFAULT_VERSION. Fixes…

### PR DESCRIPTION
**Pull request by ahmelsayed**
… #1772 

This matched this commit 7b32c189b50a26ab4289f7052c19bd12bfd11e6d

### CODE REVIEW

File: Kudu.Core/Scripts/selectNodeVersion.js [hunk 1]
```diff
@@ -34,13 +34,26 @@ function resolveNpmPath(npmRootPath, npmVersion) {
...
-        return;
+    var appSettingNpmVersion = process.env.WEBSITE_NPM_DEFAULT_VERSION;
+
+    // extract the current node version from the path
+    var nodePathSplited = nodeVersionPath.split(path.sep).filter(function(e) { return e; });
+    var currentNodeVersion = nodePathSplited[nodePathSplited.length - 1];
+
+    if (appSettingNpmVersion) {
+        return appSettingNpmVersion;
+    } else if (currentNodeVersion === '4.1.2') {
```
Comment (shrimpy) [inline Kudu.Core/Scripts/selectNodeVersion.js]: we also added v4.2.1 and v5.0 and both map to NPM 3.3.9. Do we need to do similar special handling for them as well?

Comment (shrimpy) [inline Kudu.Core/Scripts/selectNodeVersion.js]: take my comment back. npm version will be in npm.txt file. else case should take care of it.

Comment (ahmelsayed) [inline Kudu.Core/Scripts/selectNodeVersion.js]: If they already map to a 3.x npm, then no. The problem with 4.1.2 is that it was the default node version for sites, but it maps to some 2.x version of NPM. And a vanilla ASP.NET 5 app was running into path length issue with that, so we force npm to be 3.3.6

File: Kudu.Core/Scripts/selectNodeVersion.js [hunk 2]
```diff
@@ -34,13 +34,26 @@ function resolveNpmPath(npmRootPath, npmVersion) {
...
 
 function getDefaultNpmVersion(nodeVersionPath) {
-    var npmLinkPath = path.resolve(nodeVersionPath, 'npm.txt');
-    // Determine if there's a link to npm at the node path
-    if (!existsSync(npmLinkPath)) {
-        return;
+    var appSettingNpmVersion = process.env.WEBSITE_NPM_DEFAULT_VERSION;
+
+    // extract the current node version from the path
+    var nodePathSplited = nodeVersionPath.split(path.sep).filter(function(e) { return e; });
```
Comment (shrimpy) [inline Kudu.Core/Scripts/selectNodeVersion.js]: is it necessary to have the "filter" statement?

Comment (ahmelsayed) [inline Kudu.Core/Scripts/selectNodeVersion.js]: I wanted to filter out empty split results. If the path ends in `\` then the last element in the array will be an empty string. I'm not sure if the path here has a trailing slash or not though!

File: Kudu.Core/Scripts/selectNodeVersion.js [hunk 3]
```diff
@@ -34,13 +34,26 @@ function resolveNpmPath(npmRootPath, npmVersion) {
...
+    if (appSettingNpmVersion) {
+        return appSettingNpmVersion;
+    } else if (currentNodeVersion === '4.1.2') {
+        // This is to preserve parity with kudu's behavior to fix issues with ASP.NET 5 use of npm
+        return '3.3.6';
+    } else {
+        var npmLinkPath = path.resolve(nodeVersionPath, 'npm.txt');
+        // Determine if there's a link to npm at the node path
+        if (!existsSync(npmLinkPath)) {
+            return;
```
Comment (shrimpy) [inline Kudu.Core/Scripts/selectNodeVersion.js]: this function is expected to return a string. but here we return void. Shall we return null instead? or throw error?

File: Kudu.Core/Scripts/selectNodeVersion.js (hunk 2 above)
Comment (shrimpy) [inline Kudu.Core/Scripts/selectNodeVersion.js]: i see.

File: Kudu.Core/Scripts/selectNodeVersion.js (hunk 3 above)
Comment (ahmelsayed) [inline Kudu.Core/Scripts/selectNodeVersion.js]: good question. let me take a look

Comment (ahmelsayed) [inline Kudu.Core/Scripts/selectNodeVersion.js]: Updated to return null. It was there originally and the only place it's being used is checking for null/undefined, but I agree that probably returning null is the right thing to do!

Comment (shrimpy) [inline Kudu.Core/Scripts/selectNodeVersion.js]: k. looks good to me.

**Comment by suwatch** [comment]
I wonder if we can get to a point where we don't need this selectNodeVersion.js script and deploy.cmd simply relies on whatever is in the path.   The path will be determined at only one code path say in NodeJSBuilder.

**Comment by davidebbo** [comment]
@suwatch doing this in `NodeJSBuilder` is not possible, as that would break any custom deployment script. It needs to either happen for all deployments, regardless of Node (may be tricky), or needs to exist inside the script. BTW, note that despite its name, `selectNodeVersion` also generates the web.config if there isn't one.

**Pull request merged**